### ISSUE TITLE
Autofocus and Animations conflict

### ISSUE THREAD

**Issue by pronebird** [Microsoft/reactxp#526]
Hi,

I have a simple transition container that animates between old and new children. The container has two `Animated.Views` inside for each set of children. 

Since it's like a stack, the children are actually being moved between `Animated.Views`.

In order to avoid re-mount I assigned a unique key for each children to let React know that it only needs to reorder the components. i.e:

```
<View style={{flex:1}}>
  <Animated.View key={ (previousChildren ? previousChildren.key : 'null') } style={[baseStyle, previousChildrenAnimation]}>
    {previousChildren}
  </Animated.View>
  <Animated.View key={ newChildren.key } style={[baseStyle, newChildrenAnimation]}>
    {newChildren}
  </Animated.View>
</View>
```

However it seems that something fundamentally wrong happens inside `Animated.View` when its being reused. Suddenly the use of `autoFocus={ true }` inside `children` caused animations to break, the jitter can be observed if duration increased, like there is something fighting against the timing function. I briefly looked at the source of `Animated.tsx` but I couldn't figure out if it handles prop updates properly. It seems it does but there are lots of moving and it's hard to tell without a good knowledge of RX internals.

I dropped `key` from each of `Animated.Views` and everything is fine now, except that our components are being re-mounted which is not desired.

**Comment by erictraut** [comment]
This strategy and code snippet look fine (except that you haven't used RX.Styles.createViewStyle to allocate the style you're passing to the outermost View).

Note that RX.View already supports props "animateChildEnter", "animateChildLeave" and "animateChildMove". If the default animations work for your usage, you can avoid implementing this yourself. If you want a custom animation effect, you could look at how RX.View implements these features.

**Comment by pronebird** [comment]
Thanks for taking time reading this. So here the problem: for whatever mysterious reason if the component we animate has `autoFocus={ true }`, we get a random jitter and animations seem to break apart.

That's out animation container for transitions:
https://github.com/mullvad/mullvadvpn-app/blob/xp-transistions/app/components/TransitionContainer.js#L187

PS: thanks for the pointer regarding animateChildEnter/Leave. I'll take a look.

**Comment by erictraut** [comment]
I assume that previousChildren and newChildren both contain a TextInput since that's the only ReactXP component that accepts an ```autoFocus``` prop.

Are you sure that you want to use ```autoFocus```? The new TextInput will receive focus even if the previous one didn't have focus. That's because the ReactXP implementation of TextInput always calls ```focus()``` at mount time when ```autoFocus``` is true.

I don't know why calling ```focus()``` would cause an animation jitter. It may be a browser-specific behavior. Perhaps you could try on other browsers to confirm this theory.

One workaround is to manually focus the new TextInput once the animation is complete. I'd need to understand the problem more deeply to offer other potential solutions.

**Comment by pronebird** [comment]
Only `newChildren` contains `TextInput`. 

We change entire screens (it's an electron app). 

I don't know why `autoFocus` causes this, but it only happens if we use `key` on `Animated.Views`. If there are no `keys` React apparently mounts new instances and everything seems to be alright regardless `autoFocus`.

I'll make a screencast to demo what I observe.

**Comment by pronebird** [comment]
I've recorded a screencast in which you can see the push/pop animations implementation first working fine and then the second transitions (identical) which brakes because of `autoFocus={ true }`. Once I remove `autoFocus` everything works fine.

This must be super trivial implementation of transitions and I'd be surprised if we had something cheesy going on somewhere else in the app. Worst case I'll extract that work into a separate project to be able to reproduce.

https://www.dropbox.com/s/ypsfjctjsm840jd/RXAnimationsIssue.mov?dl=0

**Comment by pronebird** [comment]
Hmm weird but actually removing one of Animated.Views that does not have any children also causes jitter.

```
    return (
      <View style={{flex:1}}>
        { previousChildren &&
          (<Animated.View style={[this.state.animationStyles.style, previousChildrenAnimation]}>
            { previousChildren }
          </Animated.View>) /* only render when we have old children */ }

        <Animated.View style={[this.state.animationStyles.style, childrenAnimation]}>
          { children }
        </Animated.View>
      </View>
    );
```

**Comment by pronebird** [comment]
update:

I drop one of `Animated.Views` now after transition and the same effect can be observed regardless the presence of `key`

```
<View style={{flex:1}}>
  { /* render only if we have previousChildren */ } 
  { previousChildren  &&<Animated.View key={ (previousChildren ? previousChildren.key : 'null') } style={[baseStyle, previousChildrenAnimation]}>
    {previousChildren}
  </Animated.View> }

  <Animated.View key={ newChildren.key } style={[baseStyle, newChildrenAnimation]}>
    {newChildren}
  </Animated.View>
</View>
```

**Comment by erictraut** [comment]
No, I don't think the browser is forcing the focused input into the viewport. There's something else going on here. I suspect that you're doing something that's causing a re-render in the middle of the animation — a setState or prop update. Such re-renders will interfere with the animation because it will update the style attributes on the DOM element that's being animated.

One technique I like to use when diagnosing animation problems is to slow down the animation (say, by 10x). This will make the problem more observable and give clues as to the source of any timing-related issues.

You may also want to add a console.log() to the render() method to see if my theory is correct about re-rendering.

**Comment by pronebird** [comment]
Alright, I thought that's the case too. I was actually afraid that we received the same props twice and ran one animation over the other but that wasn't the case either.

I've recorded a little screencast here to display what happens in slow motion, and that's indeed strange and feels like we're running two animators over the same node, but that does not seem to be possible.

https://www.dropbox.com/s/hilgz96hu2pkxp5/LoggingTransitions.mov?dl=0

**Comment by erictraut** [comment]
Hmm, strange. 

I tried to replicate the problem by adding a TextInput box with autoFocus to the second panel of the "hello-world" ReactXP sample app. I couldn't replicate the problem you're seeing. It animates smoothly. This sample uses the reactxp-navigator extension for transitions, but this extension uses an animation technique similar to what you're doing in your code.

**Comment by pronebird** [comment]
I'll try to reproduce that with hello-world app. 

After a little bit of testing in slow motion what I found looked like a browser intervention indeed, some of views that contained TextInputs were simply rendered in a wrong place (i.e the layout properties of that view contradicted the visual appearance), then all properties on such views were frozen too, so changing any property via Web Inspector was impossible, except I could remove the DIV and re-add it with undo hotkey, they it would become responsive to prop changes.

The fun fact: this never occurred to me using raw CSS transitions, I am not sure what we did to piss webkit so much, unbelievable.

**Comment by erictraut** [comment]
Interesting.

I didn't realize you were using a WebKit browser. I assumed Chromium. I confirmed that the hello-world test also works correctly on Safari as well as Chrome.

Incidentally, the ReactXP uses CSS transitions to implement animations on the web. See [this file](https://github.com/Microsoft/reactxp/blob/master/src/web/animated/executeTransition.ts) if you're interested in the code. So I can't explain why it would be any different from using raw CSS transitions.

**Comment by pronebird** [comment]
Well yes, sorry Electron has Blink under the hood, which is based on Webkit AFAIK. 

Can't explain either why that happens and why it didn't happen in past.

I just replaced the push transition with CSS transition and it has exactly the same jitter as with `Animated.View`:

Steps:

1. Replaced `Animated.View` with `View`.
2. Used `setState` to reset props to `transition: none; transform: INITIAL;`
3. `setTimeout(.., 0)` and do another `setState` with `transition: transform 5s ease; transform: FINAL;`

Basically if view has autoFocus it breaks the animation. Sometimes there is this weird jitter, sometimes the transition simply breaks at once. 🤔

**Comment by pronebird** [comment]
Closing because I think it's an issue with browser. I can only guess why it used to work but most likely due to a combination of time delays etc.. especially given that we use `setState` and such that may skip run loops, cause re-render and so on.

There is a bleeding edge feature to prevent scroll however is not available on Electron 1.8.2:

```
element.focus({ preventScroll: true })
```

**Comment by lusor97** [comment]
Same problem here, I solved it creating my own autofocus function and calling it after certain ms with setTimeout
